File: android/native/emarsys/src/main/java/com/emarsys/maui/DotnetEmarsysInbox.java
package com.emarsys.maui;

import androidx.annotation.NonNull;
import com.emarsys.Emarsys;
import com.emarsys.mobileengage.api.action.ActionModel;
import com.emarsys.mobileengage.api.action.AppEventActionModel;
import com.emarsys.mobileengage.api.action.CustomEventActionModel;
import com.emarsys.mobileengage.api.action.OpenExternalUrlActionModel;
import com.emarsys.mobileengage.api.inbox.Message;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class DotnetEmarsysInbox {

    public interface FetchMessagesResultCallback {
        void onResult(List<Map<String, Object>> messages, Throwable error);
    }

    public static void fetchMessages(@NonNull FetchMessagesResultCallback resultCallback) {
        Emarsys.getMessageInbox().fetchMessages(result -> {
            if (result.getErrorCause() != null) {
                resultCallback.onResult(null, result.getErrorCause());
            } else {
                if (result.getResult() != null) {
                    List<Map<String, Object>> mappedMessages = mapMessages(result.getResult().getMessages());
                    resultCallback.onResult(mappedMessages, null);
                } else {
                    resultCallback.onResult(List.of(), null);
                }
            }
        });
    }

    public static void addTag(@NonNull String tag, @NonNull String messageId, @NonNull CompletionListener completionListener) {
        Emarsys.getMessageInbox().addTag(tag, messageId, completionListener::onCompleted);
    }

    public static void removeTag(@NonNull String tag, @NonNull String messageId, @NonNull CompletionListener completionListener) {
        Emarsys.getMessageInbox().removeTag(tag, messageId, completionListener::onCompleted);
    }

    private static @NonNull List<Map<String, Object>> mapMessages(@NonNull List<Message> messages) {
        return messages.stream().map(message -> {
            Map<String, Object> resultMap = new HashMap<>();
            resultMap.put("id", message.getId());
            resultMap.put("campaignId", message.getCampaignId());
            if (message.getCollapseId() != null) {
                resultMap.put("collapseId", message.getCollapseId());
            }
            resultMap.put("title", message.getTitle());
            resultMap.put("body", message.getBody());
            if (message.getImageUrl() != null) {
                resultMap.put("imageUrl", message.getImageUrl());
            }
            resultMap.put("receivedAt", message.getReceivedAt());
            if (message.getUpdatedAt() != null) {
                resultMap.put("updatedAt", message.getUpdatedAt());
            }
            if (message.getExpiresAt() != null) {
                resultMap.put("expiresAt", message.getExpiresAt());
            }
            if (message.getTags() != null) {
                resultMap.put("tags", message.getTags());
            }
            if (message.getProperties() != null) {
                resultMap.put("properties", message.getProperties());
            }
            if (message.getActions() != null) {
                resultMap.put("actions", mapActions(message.getActions()));
            }
            return resultMap;
        }).collect(Collectors.toList());
    }

    private static @NonNull List<Map<String, Object>> mapActions(@NonNull List<ActionModel> actions) {
        return actions.stream().map(action -> {
            Map<String, Object> actionMap = new HashMap<>();
            actionMap.put("id", action.getId());
            actionMap.put("title", action.getTitle());
            actionMap.put("type", action.getType());
            if (action instanceof AppEventActionModel) {
                AppEventActionModel appEvent = (AppEventActionModel) action;
                actionMap.put("name", appEvent.getName());
                actionMap.put("payload", appEvent.getPayload());
            } else if (action instanceof CustomEventActionModel) {
                CustomEventActionModel customEvent = (CustomEventActionModel) action;
                actionMap.put("name", customEvent.getName());
                actionMap.put("payload", customEvent.getPayload());
            } else if (action instanceof OpenExternalUrlActionModel) {
                OpenExternalUrlActionModel externalUrl = (OpenExternalUrlActionModel) action;
                actionMap.put("url", externalUrl.getUrl().toString());
            }
            return actionMap;
        }).collect(Collectors.toList());
    }

}

File: ios/native/MauiEmarsys/DotnetEmarsysInbox.swift
//
//  Copyright © 2024 Emarsys. All rights reserved.
//

import Foundation
import EmarsysSDK

@objc(DotnetEmarsysInbox)
public class DotnetEmarsysInbox: NSObject {
    
    @objc
    public static func fetchMessages(_ resultCallback: @escaping ([[String: Any]]?, Error?) -> Void) {
        Emarsys.messageInbox.fetchMessages(resultBlock: { result, error in
            if error != nil {
                resultCallback(nil, error)
            } else {
                if let messages = result?.messages {
                    let mappedMessages = mapMessages(messages)
                    resultCallback(mappedMessages, nil)
                } else {
                    resultCallback([], nil)
                }
            }
        })
    }
    
    @objc
    public static func addTag(_ tag: String, messageId: String, _ completionBlock: @escaping CompletionBlock) {
        Emarsys.messageInbox.addTag(tag: tag, messageId: messageId, completionBlock: completionBlock)
    }
    
    @objc
    public static func removeTag(_ tag: String, messageId: String, _ completionBlock: @escaping CompletionBlock) {
        Emarsys.messageInbox.removeTag(tag: tag, messageId: messageId, completionBlock: completionBlock)
    }
    
    static func mapMessages(_ messages: [EMSMessage]) -> [[String: Any]] {
        let messageDicts: [[String: Any]] = messages.map { message in
            var messageDict = [String: Any]()
            messageDict["id"] = message.id
            messageDict["campaignId"] = message.campaignId
            messageDict["collapseId"] = message.collapseId
            messageDict["title"] = message.title
            messageDict["body"] = message.body
            messageDict["imageUrl"] = message.imageUrl
            messageDict["receivedAt"] = message.receivedAt
            messageDict["updatedAt"] = message.updatedAt
            messageDict["expiresAt"] = message.expiresAt
            messageDict["tags"] = message.tags
            messageDict["properties"] = message.properties
            if let actions = message.actions {
                messageDict["actions"] = mapActions(actions)
            }
            return messageDict
        }
        return messageDicts
    }
    
    static func mapActions(_ actions: [EMSActionModelProtocol]) -> [[String: Any]] {
        let actionDicts: [[String: Any]] = actions.map { action in
            var actionDict = [String: Any]()
            if let appEvent = action as? EMSAppEventActionModel {
                actionDict["id"] = appEvent.id
                actionDict["title"] = appEvent.title
                actionDict["type"] = appEvent.type
                actionDict["name"] = appEvent.name
                actionDict["payload"] = appEvent.payload
            } else if let customEvent = action as? EMSCustomEventActionModel {
                actionDict["id"] = customEvent.id
                actionDict["title"] = customEvent.title
                actionDict["type"] = customEvent.type
                actionDict["name"] = customEvent.name
                actionDict["payload"] = customEvent.payload
            } else if let externalUrl = action as? EMSOpenExternalUrlActionModel {
                actionDict["id"] = externalUrl.id
                actionDict["title"] = externalUrl.title
                actionDict["type"] = externalUrl.type
                actionDict["url"] = externalUrl.url.absoluteString
            }
            return actionDict
        }
        return actionDicts
    }
    
}
